// @upstash/context7-sdk/dist/client.js
// src/error/index.ts
var Context7Error = class extends Error {
  constructor(message) {
    super(message);
    this.name = "Context7Error";
  }
};

// src/http/index.ts
var HttpClient = class {
  baseUrl;
  headers;
  options;
  retry;
  constructor(config) {
    this.options = {
      cache: config.cache,
      signal: config.signal
    };
    this.baseUrl = config.baseUrl.replace(/\/$/, "");
    this.headers = {
      "Content-Type": "application/json",
      ...config.headers
    };
    this.retry = typeof config?.retry === "boolean" && config?.retry === false ? {
      attempts: 1,
      backoff: () => 0
    } : {
      attempts: config?.retry?.retries ?? 5,
      backoff: config?.retry?.backoff ?? ((retryCount) => Math.exp(retryCount) * 50)
    };
  }
  async request(req) {
    const method = req.method || "POST";
    let url = [this.baseUrl, ...req.path ?? []].join("/");
    if (method === "GET" && req.query) {
      const queryParams = new URLSearchParams();
      Object.entries(req.query).forEach(([key, value]) => {
        if (value !== void 0) {
          queryParams.append(key, String(value));
        }
      });
      const queryString = queryParams.toString();
      if (queryString) {
        url += `?${queryString}`;
      }
    }
    const requestOptions = {
      cache: this.options.cache,
      method,
      headers: this.headers,
      body: req.body ? JSON.stringify(req.body) : void 0,
      keepalive: true,
      signal: this.options.signal?.()
    };
    let res = null;
    let error = null;
    for (let i = 0; i <= this.retry.attempts; i++) {
      try {
        res = await fetch(url, requestOptions);
        break;
      } catch (error_) {
        if (requestOptions.signal?.aborted) {
          throw error_;
        }
        error = error_;
        if (i < this.retry.attempts) {
          await new Promise((r) => setTimeout(r, this.retry.backoff(i)));
        }
      }
    }
    if (!res) {
      throw error ?? new Error("Exhausted all retries");
    }
    if (!res.ok) {
      const errorBody = await res.json();
      throw new Context7Error(errorBody.error || errorBody.message || res.statusText);
    }
    const contentType = res.headers.get("content-type");
    if (contentType?.includes("application/json")) {
      const body = await res.json();
      return { result: body };
    } else {
      const text = await res.text();
      const headers = this.extractTxtResponseHeaders(res.headers);
      return { result: text, headers };
    }
  }
  extractTxtResponseHeaders(headers) {
    const page = headers.get("x-context7-page");
    const limit = headers.get("x-context7-limit");
    const totalPages = headers.get("x-context7-total-pages");
    const hasNext = headers.get("x-context7-has-next");
    const hasPrev = headers.get("x-context7-has-prev");
    const totalTokens = headers.get("x-context7-total-tokens");
    if (!page || !limit || !totalPages || !hasNext || !hasPrev || !totalTokens) {
      return void 0;
    }
    return {
      page: parseInt(page, 10),
      limit: parseInt(limit, 10),
      totalPages: parseInt(totalPages, 10),
      hasNext: hasNext === "true",
      hasPrev: hasPrev === "true",
      totalTokens: parseInt(totalTokens, 10)
    };
  }
};

// src/commands/command.ts
var Command = class {
  request;
  endpoint;
  constructor(request, endpoint) {
    this.request = request;
    this.endpoint = endpoint;
  }
  /**
   * Execute the command using a client.
   */
  async exec(client) {
    const { result } = await client.request({
      method: this.request.method || "POST",
      path: [this.endpoint],
      query: this.request.query,
      body: this.request.body
    });
    if (result === void 0) {
      throw new TypeError("Request did not return a result");
    }
    return result;
  }
};

// src/utils/format.ts
function formatCodeSnippet(snippet) {
  const codeBlocks = snippet.codeList.map((c) => `\`\`\`${c.language}
${c.code}
\`\`\``).join("\n\n");
  const content = snippet.codeDescription ? `${snippet.codeDescription}

${codeBlocks}` : codeBlocks;
  return {
    title: snippet.codeTitle,
    content,
    source: snippet.codeId
  };
}
function formatInfoSnippet(snippet) {
  return {
    title: snippet.breadcrumb || "Documentation",
    content: snippet.content,
    source: snippet.pageId
  };
}
function formatLibrary(r) {
  return {
    id: r.id,
    name: r.title,
    description: r.description,
    totalSnippets: r.totalSnippets ?? 0,
    trustScore: r.trustScore ?? 0,
    benchmarkScore: r.benchmarkScore ?? 0,
    versions: r.versions
  };
}
function getTrustScoreLabel(trustScore) {
  if (trustScore === void 0 || trustScore < 0) return "Unknown";
  if (trustScore >= 7) return "High";
  if (trustScore >= 4) return "Medium";
  return "Low";
}
function formatLibraryAsText(library) {
  const lines = [
    `- Title: ${library.name}`,
    `- Context7-compatible library ID: ${library.id}`,
    `- Description: ${library.description}`
  ];
  if (library.totalSnippets > 0) {
    lines.push(`- Code Snippets: ${library.totalSnippets}`);
  }
  lines.push(`- Trust Score: ${getTrustScoreLabel(library.trustScore)}`);
  if (library.benchmarkScore > 0) {
    lines.push(`- Benchmark Score: ${library.benchmarkScore}`);
  }
  if (library.versions && library.versions.length > 0) {
    lines.push(`- Versions: ${library.versions.join(", ")}`);
  }
  return lines.join("\n");
}
function formatLibrariesAsText(libraries) {
  if (libraries.length === 0) {
    return "No documentation libraries found matching your query.";
  }
  return libraries.map(formatLibraryAsText).join("\n----------\n");
}

// src/commands/get-context/index.ts
var DEFAULT_TYPE = "json";
var GetContextCommand = class extends Command {
  responseType;
  constructor(query, libraryId, options) {
    const queryParams = {};
    queryParams.query = query;
    queryParams.libraryId = libraryId;
    const responseType = options?.type ?? DEFAULT_TYPE;
    queryParams.type = responseType;
    super({ method: "GET", query: queryParams }, "v2/context");
    this.responseType = responseType;
  }
  async exec(client) {
    const { result } = await client.request({
      method: this.request.method || "GET",
      path: [this.endpoint],
      query: this.request.query,
      body: this.request.body
    });
    if (result === void 0) {
      throw new Context7Error("Request did not return a result");
    }
    if (this.responseType === "txt" && typeof result === "string") {
      return result;
    }
    const apiResult = result;
    const codeDocs = apiResult.codeSnippets.map(formatCodeSnippet);
    const infoDocs = apiResult.infoSnippets.map(formatInfoSnippet);
    return [...codeDocs, ...infoDocs];
  }
};

// src/commands/search-library/index.ts
var DEFAULT_TYPE2 = "json";
var SearchLibraryCommand = class extends Command {
  responseType;
  constructor(query, libraryName, options) {
    if (!query || !libraryName) {
      throw new Context7Error("query and libraryName are required");
    }
    const queryParams = {};
    queryParams.query = query;
    queryParams.libraryName = libraryName;
    super({ method: "GET", query: queryParams }, "v2/libs/search");
    this.responseType = options?.type ?? DEFAULT_TYPE2;
  }
  async exec(client) {
    const { result } = await client.request({
      method: this.request.method || "GET",
      path: [this.endpoint],
      query: this.request.query
    });
    if (result === void 0) {
      throw new Context7Error("Request did not return a result");
    }
    const libraries = result.results.map(formatLibrary);
    if (this.responseType === "txt") {
      return formatLibrariesAsText(libraries);
    }
    return libraries;
  }
};

// src/client.ts
var DEFAULT_BASE_URL = "https://context7.com/api";
var API_KEY_PREFIX = "ctx7sk";
var Context7 = class {
  httpClient;
  constructor(config = {}) {
    const apiKey = config.apiKey || process.env.CONTEXT7_API_KEY;
    if (!apiKey) {
      throw new Context7Error(
        "API key is required. Pass it in the config or set CONTEXT7_API_KEY environment variable."
      );
    }
    if (!apiKey.startsWith(API_KEY_PREFIX)) {
      console.warn(`API key should start with '${API_KEY_PREFIX}'`);
    }
    this.httpClient = new HttpClient({
      baseUrl: DEFAULT_BASE_URL,
      headers: {
        Authorization: `Bearer ${apiKey}`
      },
      retry: {
        retries: 5,
        backoff: (retryCount) => Math.exp(retryCount) * 50
      },
      cache: "no-store"
    });
  }
  /**
   * Search for libraries matching the given query
   * @param query The user's question or task (used for relevance ranking)
   * @param libraryName The library name to search for
   * @param options Response format options
   * @returns Array of matching libraries (json) or formatted text (txt)
   */
  async searchLibrary(query, libraryName, options) {
    const command = new SearchLibraryCommand(query, libraryName, options);
    return await command.exec(this.httpClient);
  }
  /**
   * Get documentation context for a library
   * @param query The user's question or task
   * @param libraryId Context7 library ID (e.g., "/facebook/react")
   * @param options Response format options
   * @returns Documentation as Documentation[] (json, default) or string (txt)
   */
  async getContext(query, libraryId, options) {
    const command = new GetContextCommand(query, libraryId, options);
    return await command.exec(this.httpClient);
  }
};
export {
  Context7,
  Context7Error
};
